sklearn/metrics/_pairwise_distances_reduction/_argkminlabels.pyx.tp

```diff
@@ -0,0 +1,220 @@
+{{py:
+
+implementation_specific_values = [
+    # Values are the following ones:
+    #
+    #       name_suffix, INPUT_DTYPE_t, INPUT_DTYPE
+    #
+    # We also use the float64 dtype and C-type names as defined in
+    # `sklearn.utils._typedefs` to maintain consistency.
+    #
+    ('64', 'DTYPE_t', 'DTYPE'),
+    ('32', 'cnp.float32_t', 'np.float32')
+]
+
+}}
+
+from cython cimport floating, integral
+from cython.parallel cimport parallel, prange
+from libcpp.map cimport map as cpp_map, pair as cpp_pair
+from libc.stdlib cimport free
+
+cimport numpy as cnp
+
+cnp.import_array()
+
+from ...utils._typedefs cimport ITYPE_t, DTYPE_t
+from ...utils._typedefs import ITYPE, DTYPE
+from ...utils._sorting cimport simultaneous_sort
+import numpy as np
+from scipy.sparse import issparse
+from sklearn.utils.fixes import threadpool_limits
+
+cpdef enum WeightingStrategy:
+    uniform = 0
+    distance = 1
+    other = 2
+
+{{for name_suffix, INPUT_DTYPE_t, INPUT_DTYPE in implementation_specific_values}}
+from ._argkmin cimport ArgKmin{{name_suffix}}, EuclideanArgKmin{{name_suffix}}
+from ._datasets_pair cimport DatasetsPair{{name_suffix}}
+
+cdef class ArgKminLabels{{name_suffix}}(ArgKmin{{name_suffix}}):
```

Comment:
I think we can find a more descriptive name for this reducer. What about: `ArgKminDistanceWeightedClassMode`?

It's a bit long but since it's not part of the public API I would rather go towards being explicit. I would also consider:

- `ArgKminWeightedClassMode` (probably a good balance between concise and precise naming;
- `ArgKminWeightedLabelMode`.

I would rather use "Class" instead of "Label" to better highlight that this is useful for (multiclass) classification models. Futhermore this code expects with consecutive integer class indices between 0 and n_classes - 1 and does not accepts the raw class labels in `y` passed to fit (which can be non-ordered, non-integer labels such as `str` objects, pandas categories) before we do the typical idiom in `classifier.fit`:

```python
            label_encoder_ = LabelEncoder().fit(y)
            self.classes_ = label_encoder_.classes_
            n_classes = len(self.classes_)
```

which by-the way we should probably factorize at some point in scikit-learn :)

Reply:
Actually for nearest neighbor classification, the class label to class index mapping is more involved because this estimator is one of the few that supports multioutput multiclass `y` values:

- https://github.com/scikit-learn/scikit-learn/pull/24076/files#diff-945a1221f914501203afb06a3bc50d8bff40dcd6fd29571ad646267001e27697L477-L481

But the rest of the argument about class (index) vs (class) label naming is still valid.

Reply:
I can agree with the logic and urge to favor being explicit, however the proposed names would make the _file names_ pretty difficult. What do you think about something like `ArgKminClassMode` with the "weighted" being implied and explained in documentation. Its corresponding file name could be `_argkminclassmode.pyx.tp` which is still manageable imo

Reply:
+1 for `ArgKminClassMode`.

Reply:
https://github.com/Micky774/scikit-learn/pull/10 proposes this renaming.